storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateEntry.java

```diff
@@ -103,28 +121,44 @@ public boolean maybeUpdateProducerEpoch(short producerEpoch) {
             return false;
         }
     }
+    
+    public void maybeUpdateTentaitiveSequence(int sequence) {
+        if (batchMetadata.isEmpty() && (!this.tentativeSequence.isPresent() || this.tentativeSequence.getAsInt() > sequence))
+            this.tentativeSequence = OptionalInt.of(sequence);
+    }
 
     private void addBatchMetadata(BatchMetadata batch) {
+        // When appending a batch, we no longer need tentative sequence.
+        this.tentativeSequence = OptionalInt.empty();
         if (batchMetadata.size() == ProducerStateEntry.NUM_BATCHES_TO_RETAIN) batchMetadata.removeFirst();
         batchMetadata.add(batch);
     }
+    
+    public boolean compareAndSetVerificationState(short expectedProducerEpoch, VerificationState expectedVerificationState, VerificationState newVerificationState) {
+        if (expectedProducerEpoch == this.producerEpoch && verificationState == expectedVerificationState) {
+            this.verificationState = newVerificationState;
+            return true;
+        }
+        return false;
+    }
 
     public void update(ProducerStateEntry nextEntry) {
-        update(nextEntry.producerEpoch, nextEntry.coordinatorEpoch, nextEntry.lastTimestamp, nextEntry.batchMetadata, nextEntry.currentTxnFirstOffset);
+        update(nextEntry.producerEpoch, nextEntry.coordinatorEpoch, nextEntry.lastTimestamp, nextEntry.batchMetadata, nextEntry.currentTxnFirstOffset, nextEntry.verificationState);
     }
 
     public void update(short producerEpoch, int coordinatorEpoch, long lastTimestamp) {
-        update(producerEpoch, coordinatorEpoch, lastTimestamp, new ArrayDeque<>(0), OptionalLong.empty());
+        update(producerEpoch, coordinatorEpoch, lastTimestamp, new ArrayDeque<>(0), OptionalLong.empty(), VerificationState.EMPTY);
     }
 
     private void update(short producerEpoch, int coordinatorEpoch, long lastTimestamp, Deque<BatchMetadata> batchMetadata,
-                        OptionalLong currentTxnFirstOffset) {
+                        OptionalLong currentTxnFirstOffset, VerificationState verificationState) {
```

Comment:
nit: when arg lists go above 2 or 3, it's helpful to start putting each argument on a separate line:
```java
private void update(
  short producerEpoch,
  int coordinatorEpoch,
  long lastTimestamp,
  ...
) {
````